src/Components/Components/src/NavigationManager.cs

```diff
@@ -274,6 +281,111 @@ protected void NotifyLocationChanged(bool isInterceptedLink)
         }
     }
 
+    /// <summary>
+    /// Notifies the registered handlers of the current location change.
+    /// </summary>
+    /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI.</param>
+    /// <param name="isNavigationIntercepted">Whether this navigation was intercepted from a link.</param>
+    /// <returns>A <see cref="ValueTask{TResult}"/> representing the completion of the operation. If the result is <see langword="true"/>, the navigation should continue.</returns>
+    protected async ValueTask<bool> NotifyLocationChangingAsync(string uri, bool isNavigationIntercepted)
+    {
+        _locationChangingCts?.Cancel();
+        _locationChangingCts = null;
+
+        var handlerCount = _locationChangingHandlers.Count;
+
+        if (handlerCount == 0)
+        {
+            return true;
+        }
+
+        _locationChangingCts = new();
+
+        var context = new LocationChangingContext(uri, isNavigationIntercepted, _locationChangingCts);
+        var cancellationToken = _locationChangingCts.Token;
+
+        try
+        {
+            if (handlerCount == 1)
+            {
+                await _locationChangingHandlers[0].Invoke(context);
+            }
+            else
+            {
+                var locationChangingHandlersCopy = ArrayPool<LocationChangingHandler>.Shared.Rent(handlerCount);
+
+                try
+                {
+                    _locationChangingHandlers.CopyTo(locationChangingHandlersCopy);
+
+                    var locationChangingTasks = new Task[handlerCount];
+
+                    for (var i = 0; i < handlerCount; i++)
+                    {
+                        locationChangingTasks[i] = locationChangingHandlersCopy[i].Invoke(context);
+
+                        if (cancellationToken.IsCancellationRequested)
+                        {
+                            return false;
+                        }
+                    }
+
+                    await Task.WhenAll(locationChangingTasks).WaitAsync(cancellationToken);
+                }
+                finally
+                {
+                    ArrayPool<LocationChangingHandler>.Shared.Return(locationChangingHandlersCopy);
+                }
+            }
+
+            return !cancellationToken.IsCancellationRequested;
+        }
+        catch (TaskCanceledException ex) when (ex.CancellationToken == cancellationToken)
```

Comment:
In other cases we've made a point of avoiding exception filters because they had poor perf/behavior characteristics on WebAssembly AOT. @lewing Do you know if that is fixed these days - are we free to use exception filters now, or is it still better to avoid them?

Example: See `ComponentBase` inside `RunInitAndSetParametersAsync`.

If it's just as easy to do the equality check inside the `catch` and manually rethrow, that might be better.